### PR TITLE
Cycle Gan

### PR DESCRIPTION
**Pull request by husonchen**
I add a script to train cycle gan.

### CODE REVIEW

**Comment by mli** [comment]
Job PR-354-1 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-354/1/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-354-3 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-354/3/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-354-4 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-354/4/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-354-5 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-354/5/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-354-6 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-354/6/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-354-7 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-354/7/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-354-8 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-354/8/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-354-10 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-354/10/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-354-14 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-354/14/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-354-15 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-354/15/index.html 
Code coverage of this PR:  vs. Master:

**Comment by zhreshold** [comment]
@husonchen  Sorry about the delayed review.
Do you still have input to convert all bash scripts into python scripts? In which case windows users may benefit from that.

**Comment by husonchen** [comment]
@zhreshold I have turned bash script to python script.

**Comment by mli** [comment]
Job PR-354-18 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-354/18/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-354-19 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-354/19/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-354-20 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-354/20/index.html 
Code coverage of this PR:  vs. Master:

File: scripts/gan/cycle_gan/README.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+## Reproducing Cycle GAN experiments
+
+
+**Download horse2zebra dataset**
+```bash
+bash ./download_dataset.sh horse2zebra
```
Comment (zhreshold) [inline scripts/gan/cycle_gan/README.md]: python download_dataset.py --file horse2zebra

File: scripts/gan/cycle_gan/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+```bash
+python train_cgan.py --dataroot ./horse2zebra
+```
+![images](images.png "images during training")
+The meaning of those images are :
+
+| | | | |
+|-|-|-|-|
+| real_A(from dataset) | fake_B(generate from real_A) | rec_A(reconstruct from fake_B) | idt_A(generate from real_B) |
+| real_B(from dataset) | fake_A(generate from real_A) | rec_B(reconstruct from fake_A) | idt_B(generate from real_A) |
```
Comment (zhreshold) [inline scripts/gan/cycle_gan/README.md]: can you add a very simple demo.py?

File: scripts/gan/cycle_gan/train_cgan.py [hunk 3]
```diff
@@ -0,0 +1,579 @@
...
+        optimizer_DB.step(opt.batchSize)
+        if ((epoch-1) * dataset_size + i) % 100 == 0:
+            plot_loss(losses_log, (epoch-1) * dataset_size + i,epoch,i)
+            plot_img(losses_log)
+    lr = (1.0 - max(0, epoch + 1 + opt.epoch_count - opt.niter) / float(opt.niter_decay + 1)) * opt.lr
+    optimizer_GA.set_learning_rate(lr)
+    optimizer_GB.set_learning_rate(lr)
+    optimizer_DA.set_learning_rate(lr)
+    optimizer_DB.set_learning_rate(lr)
+    netG_A.save_parameters('{0}/netG_A_epoch_{1}.pth'.format(opt.experiment, epoch))
```
Comment (zhreshold) [inline scripts/gan/cycle_gan/train_cgan.py]: pth -> params for consistency

**Comment by husonchen** [comment]
Demo script has been added.

**Comment by mli** [comment]
Job PR-354-22 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-354/22/index.html 
Code coverage of this PR:  vs. Master:

**Comment by zhreshold** [comment]
@hetong007 @zhanghang1989 I will merge this in 24hr if no other comment.
@husonchen Appreciated for the continuous contribution!

**Pull request merged**